library/alloc/src/boxed/thin.rs

```diff
@@ -0,0 +1,198 @@
+// Based on
+// https://github.com/matthieu-m/rfc2580/blob/b58d1d3cba0d4b5e859d3617ea2d0943aaa31329/examples/thin.rs
+// by matthieu-m
+use crate::alloc::{self, Layout};
+use core::cmp;
+use core::fmt::{self, Debug, Formatter};
+use core::marker::{PhantomData, Unsize};
+use core::mem;
+use core::ops::{Deref, DerefMut};
+use core::ptr::Pointee;
+use core::ptr::{self, NonNull};
+
+/// ThinBox.
+///
+/// A thin pointer for heap allocation, regardless of T.
+///
+/// # Examples
+///
+/// ```
+/// #![feature(thin_box)]
+/// use std::boxed::ThinBox;
+///
+/// let five = ThinBox::new(5);
+/// let thin_slice = ThinBox::<[i32]>::new_unsize([1, 2, 3, 4]);
+///
+/// use std::mem::{size_of, size_of_val};
+/// let size_of_ptr = size_of::<*const ()>();
+/// assert_eq!(size_of_ptr, size_of_val(&five));
+/// assert_eq!(size_of_ptr, size_of_val(&thin_slice));
+/// ```
+#[unstable(feature = "thin_box", issue = "92791")]
+pub struct ThinBox<T: ?Sized> {
+    ptr: WithHeader<<T as Pointee>::Metadata>,
+    _marker: PhantomData<fn(T) -> T>,
```

Comment:
I assume this `PhantomData` is there for invariance, but I think (not sure though) that using `<T as Pointee>::Metadata` is enough here. Additionally, there should be a `PhantomData<T>` for dropck.

Reply:
> but I think (not sure though) that using `<T as Pointee>::Metadata` is enough here

Can you explain how for me? I believe you I just don't understand how.

Reply:
Even if `T` can coerce to `U`, the compiler cannot guarantee that `<T as Pointee>::Metadata` can coerce to `<U as Pointee>::Metadata`, which makes the struct invariant.

[Playground example](https://play.rust-lang.org/?gist=b91fa44920e7dfcc47535d52be7c6f15)

Reply:
Okay, that makes sense, I went ahead and updated the PhantomData to T so I think this should be resolved, thank you!